src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -10,12 +10,50 @@ import scala.util.Try
 // ***************************************
 // * The union types for for the parser. *
 // ***************************************
-case class Variable(varName : String, relName : String, index : Int )
+// case class Variable(varName : String, relName : String, index : Int )
 // TODO make Atom a trait, and have multiple case classes, e.g., RelationAtom and CondExprAtom
-case class Atom(name : String, terms : List[Variable])
+// ddlog column variable type: constant or variable
+sealed trait ColumnVariable
+case class Variable(varName : String, relName : String, index : Int ) extends ColumnVariable
+
+case class Expression(variables: List[ColumnVariable], ops: List[String], relName: String, index: Int)
+case class ColumnExpr(expr: Expr, relName: String, index: Int)
+case class Operator(operator: String, operand: ColumnVariable)
+
+sealed trait Expr
+case class VarExpr(name: String) extends Expr
+case class ConstExpr(value: String) extends Expr
+case class FuncExpr(function: String, args: List[Expr], isAggregation: Boolean) extends Expr
+case class BinaryOpExpr(lhs: Expr, op: String, rhs: Expr) extends Expr
+
+case class Atom(name : String, terms : List[ColumnExpr])
 case class Attribute(name : String, terms : List[Variable], types : List[String])
-case class ConjunctiveQuery(head: Atom, bodies: List[List[Atom]])
+case class ConjunctiveQuery(head: Atom, bodies: List[List[Atom]], conditions: List[Option[Cond]], isDistinct: Boolean)
 case class Column(name : String, t : String)
+case class BodyWithCondition(body: List[Atom], condition: Option[Cond])
+
+// condition
+sealed trait Cond
+case class ComparisonCond(lhs: Expr, op: String, rhs: Expr) extends Cond
+case class NegationCond(cond: Cond) extends Cond
+case class BinaryOpCond(lhs: Cond, op: LogicOperator.LogicOperator, rhs: Cond) extends Cond
```

Comment:
BinaryCond feels like a condition whose return type is binary... probably let's go with your first proposal `CompoundCond`?